static/src/javascripts/projects/common/modules/commercial/prebid/prebid-service.js

```diff
@@ -0,0 +1,130 @@
+                    window.pbjs.que.push(function () {
+                        pbjs.addAdUnits(pbjsAdUnits);
+                        pbjs.requestBids({
+                            bidsBackHandler : function () {
+                                // We are not actually going to use the bid responses; this test purely validates performance
+                                googletag.display(slotId);
+                                resolve();
+                            }
+                        });
+                    });
+                });
+            });
+
+            return pendingRequests;
+        };
+
+    }
+
+    function PrebidAdUnit(slotElementId) {
+        this.code = slotElementId;
+        this.sizes = getSlotAdSizes(slotElementId);
+        this.bids = getAppnexusBids(this.sizes);
+    }
+
+    function getAppnexusBids(slotSizes) {
+        var slotPlacements = placements;
+
+        slotPlacements = filter(slotPlacements, function matchEdition(placement) {
+            return placement.edition === config.page.edition;
+        });
+
+        slotPlacements = filter(slotPlacements, function matchViewport(placement) {
+            return detect.isBreakpoint(placement.breakpoint);
+        });
+
+        slotPlacements = filter(slotPlacements, function matchSlotSize(placement) {
+            return find(slotSizes, function (slotSize) {
+                return slotSize[0] === placement.width && slotSize[1] === placement.height;
+            });
+        });
+
+        return map(slotPlacements, function toPrebidPlacement(placement) {
+            return {
+                bidder : 'appnexus',
+                params : {
+                    placementId : placement.id,
+                    referrer : 'http://www.theguardian.com/uk'
+                }
+            };
+        });
+    }
+
+    function getSlotAdSizes(slotElementId) {
+        var $slot = bonzo(document.getElementById(slotElementId));
+
+        var validBreakpoints = filter(detect.breakpoints, function (breakpoint) {
+            return detect.isBreakpoint({min : breakpoint.name});
+        });
+
+        var validBreakpointKeys = map(validBreakpoints, function (breakpoint) {
+            // Duplicated from dfp-api for the purpose of our test code
+            return breakpoint.name.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();
+        });
+
+        var largestValidBreakpointDefinition = findLast(validBreakpointKeys, function (key) {
+            return $slot.data(key) !== undefined;
+        });
+
+        var sizeMappingString = $slot.data(largestValidBreakpointDefinition);
+        var sizeStrings = sizeMappingString.split('|');
+        return map(sizeStrings, function (sizeString) {
+            var dimensions = sizeString.split(',');
+            return map(dimensions, function (widthOrHeightString) {
+                return parseInt(widthOrHeightString, 10);
+            });
+        });
+    }
+
+    return new PrebidTestService();
+
+});
```

Comment:
I did consider that. My thinking was that with a constructor -

- it was easy to write a stateless module, which in turn made things a bit nicer to test. It's true you can break this quite easily (as per `save-for-later`), but it's also relatively easy to fix this by moving initialization into the constructor
- you don't have to add an `init` method. You don't need a third party bootstrap script to initialize your dependency, and when you're looking at the module itself, you don't have to worry that someone might be calling `init` a second time innocently. If you want a clean module, you have to call `new service.Constructor` - in which case, it's obvious that you're doing something unusual, and you aren't going to wipe out the state some other module is depending on.
- it's a minor stylistic point, but because your 'class' definition is in a function, you can describe the public interface at the top of a file, even though your return statement is at the very bottom. But then again, it's hardly arduous to pageDown a file. You also get to declare your 'type' in a function rather than an object literal, so you don't have to mess about with joining commas. But that is fairly trivial.

I do have my own biases, though - I worked at a Java shop for many years, and so several of the conventions I picked up were chosen to make scripts look a lot more Java-esque. Conventions like always having an `init` method might just be a house style I need to embrace.